builder/build_validate.go

```diff
@@ -254,46 +263,81 @@ func (b *Builder) mcaPkgInfo(mInfo map[string]*swupd.Manifest, version, download
 	pInfo := make(map[string]*mcaBundlePkgInfo)
 
 	// Download and query file metadata from all packages in each bundle
+	bundleWorker := func() {
+		for m := range mCh {
+			var pkgList = []string{}
+			for pkg := range mInfo[m.name].BundleInfo.AllPackages {
+				pkgList = append(pkgList, pkg)
+			}
+
+			out, err := downloadRpms(packagerCmd, pkgList, buildVersionDir, downloadRetries)
+			if err != nil {
+				errorCh <- err
+				wg.Done()
+				return
+			}
+
+			// Collect metadata to resolve installed RPM file names
+			bundlePkgInfo := pkgInfoFromNoopInstall(out.String())
+			for _, p := range bundlePkgInfo {
+				rw.RLock()
+				name := pkgCache[p.name]
+				rw.RUnlock()
+
+				// Resolve files for package when it doesn't exist in the cache
+				if name == nil {
+					p.files, err = b.resolvePkgFiles(p, version)
+					if err != nil {
+						errorCh <- err
+						wg.Done()
+						return
+					}
+					rw.Lock()
+					pkgCache[p.name] = p
+					rw.Unlock()
```

Comment:
it is better to do `m.allPkgs[p.name] = p` here and keep the code below inside the else block. This way you prevent releasing the lock just to reacquire it right after. Since `p` is a local variable, the assignment can be done out of a lock (and also outside the if block i suggested above if you decide to use it).

Reply:
Re-added the else statement. Also, I removed the name variable and reduced the number of rw.RLock() calls to one.